### PR TITLE
Telemetry: add way to test what metrics would be emitted

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
Developers cannot easily test telemetry metrics emitted during test execution.

## Solution
Add 'logger' to keep track of recorded metrics. Lots of rough edges still since our telemetry system maps all metadata to strings...
Though I guess it makes sense to leave it as-is because the logger is storing what will actually be sent to out vs. what the module code sent in. Could also add some temporary logic to intercept metrics prior to being mapped, or perhaps just move the mapping logic.

## TODO:
This PR is just the first half for better telemetry inspection. The next half is hooking up the code to Winston.
In release the logger will just use the current behavior, but in dev it'll output more info.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/telemetry/defaultTelemetryClient.ts [hunk 1]
```diff
@@ -18,13 +18,22 @@ import { DefaultSettingsConfiguration } from '../settingsConfiguration'
...
-    public static readonly DEFAULT_IDENTITY_POOL = 'us-east-1:820fd6d1-95c0-4ca4-bffb-3f01d32da842'
-    public static readonly DEFAULT_TELEMETRY_ENDPOINT = 'https://client-telemetry.us-east-1.amazonaws.com'
-
+    private static readonly DEFAULT_IDENTITY_POOL = 'us-east-1:820fd6d1-95c0-4ca4-bffb-3f01d32da842'
+    private static readonly DEFAULT_TELEMETRY_ENDPOINT = 'https://client-telemetry.us-east-1.amazonaws.com'
     private static readonly PRODUCT_NAME = 'AWS Toolkit For VS Code'
+    private static endpoint: string
+    public static identityPool: string
+
+    static {
```
Comment (JadenSimon) [inline src/shared/telemetry/defaultTelemetryClient.ts]: Can't use static blocks I guess.
Need a better way to load-in these dev settings without adding to the prototype. Another PR I suppose.

File: src/shared/telemetry/defaultTelemetryService.ts [hunk 2]
```diff
@@ -151,20 +166,17 @@ export class DefaultTelemetryService implements TelemetryService {
...
-            async () => {
-                await this.flushRecords()
-                this._timer?.refresh()
-            },
-            this._flushPeriod
-        )
+    private startTimer(): void {
+        // The timer is not reset until after the flush as completed
+        this._timer = globals.clock.setTimeout(async () => {
+            await this.flushRecords()
```
Comment (JadenSimon) [inline src/shared/telemetry/defaultTelemetryService.ts:168]: Fake timers do not await callbacks, so need to fix this at some point.

File: src/test/shared/credentials/loginManager.test.ts [hunk 3]
```diff
@@ -66,7 +65,18 @@ describe('LoginManager', async function () {
     })
 
     function assertTelemetry(expected: Parameters<typeof recordAwsValidateCredentials>[0]): void {
-        assert.deepStrictEqual(recordAwsValidateCredentialsSpy.args[0][0], expected)
+        const passive = expected.passive
```
Comment (JadenSimon) [inline src/test/shared/credentials/loginManager.test.ts:68]: Passive is outside of metadata which is unfortunate, though makes sense as it is attached to all metrics. So I'll need to make a better way to test for it.

File: src/shared/telemetry/defaultTelemetryService.ts (hunk 2 above)
Comment (justinmk3) [reply src/shared/telemetry/defaultTelemetryService.ts:168]: Fake timers continue to be a burden. Perhaps it makes sense to make the timer values globally-controllable, so that tests can control them. Then real timers are used, but with adjusted times.

File: src/shared/logger/loggableType.ts [hunk 4]
```diff
@@ -1,9 +0,0 @@
-/*!
- * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
- * SPDX-License-Identifier: Apache-2.0
- */
-
-/**
- * Loggable types: Error, string
- */
-export type Loggable = Error | string
```
Comment (justinmk3) [inline src/shared/logger/loggableType.ts:9]: thank you :D

File: src/shared/settingsConfiguration.ts [hunk 5]
```diff
@@ -16,7 +16,13 @@ import globals from './extensionGlobals'
  */
 export type SettingsConfiguration = ClassToInterfaceType<DefaultSettingsConfiguration>
 
-export type AwsDevSetting = 'aws.forceCloud9' | 'aws.dev.forceTelemetry' | 'aws.dev.forceInstallTools' | 'aws.dev.foo'
+export type AwsDevSetting =
+    | 'aws.forceCloud9'
+    | 'aws.dev.forceTelemetry'
+    | 'aws.dev.forceInstallTools'
+    | 'aws.dev.foo'
```
Comment (justinmk3) [inline src/shared/settingsConfiguration.ts:23]: we don't need "foo" anymore, since we have plenty of real examples now :)

**Comment by JadenSimon** [comment]
Winston can indeed query logs by field so it could be used to store metrics instead of this ad-hoc implementation. Logic to filter/map would still be needed though.

File: src/shared/telemetry/defaultTelemetryService.ts (hunk 2 above)
Comment (JadenSimon) [reply src/shared/telemetry/defaultTelemetryService.ts:168]: Yup agreed. I'm thinking of ways to move away from it. The problem with real timers is that they are very inaccurate. So test code would need to account for this :/

File: src/shared/telemetry/telemetryLogger.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+interface MetricQuery extends MetadataQuery {
+    readonly returnMetric: true
+}
+
+interface Metadata {
+    [key: string]: string | undefined
+}
+
+export type Query = MetricQuery | MetadataQuery
```
Comment (bryceitoc9) [inline src/shared/telemetry/telemetryLogger.ts:25]: Unnecessary export. Is the type even necessary?

File: src/shared/telemetry/telemetryLogger.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        return datum.Key !== undefined && datum.Value !== undefined
+    }
+
+    /**
+     * Telemetry currently sends metadata as an array of key/value pairs, but this is unintuitive for JS
+     */
+    private mapMetadata(metadata: Required<MetricDatum>['Metadata'], filters: string[]): Metadata {
+        const result: Metadata = {}
+        return metadata
+            .filter(this.isValidEntry.bind(this))
```
Comment (bryceitoc9) [inline src/shared/telemetry/telemetryLogger.ts:76]: Why is `.bind(this)` necessary? Can you just arrow function your param in?

File: src/shared/telemetry/telemetryLogger.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    public query(query: MetadataQuery): Metadata[]
+    public query(query: Query): Metadata[] | MetricDatum[] {
+        const metrics = this._metrics.filter(m => m.MetricName === query.metricName)
+
+        if (!query.returnMetric) {
+            return metrics.map(m => m.Metadata ?? []).map(m => this.mapMetadata(m, query.filters ?? []))
+        }
+
+        return metrics
+    }
```
Comment (bryceitoc9) [inline src/shared/telemetry/telemetryLogger.ts:60]: I'm not huge on the one-function-with-multiple-return-types-and-constructors. Looks like you're only using the no-`selectMetadata` case in one place...can you move the specific logic to that, or just break it out into another function? Is this even necessary for what you're trying to do there?

File: src/shared/telemetry/telemetryLogger.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    public reset(): void {
```
Comment (bryceitoc9) [inline src/shared/telemetry/telemetryLogger.ts:62]: This is not being called.

File: src/shared/telemetry/defaultTelemetryService.ts [hunk 10]
```diff
@@ -131,11 +138,11 @@ export class DefaultTelemetryService implements TelemetryService {
             const eventWithCommonMetadata = this.injectCommonMetadata(event, actualAwsContext)
             this._eventQueue.push(eventWithCommonMetadata)
             getLogger().debug(`telemetry: passive=${event.Passive ? 1 : 0} ${event.MetricName}`)
-        }
-    }
 
-    public get records(): ReadonlyArray<MetricDatum> {
-        return this._eventQueue
+            if (!isReleaseVersion()) {
```
Comment (bryceitoc9) [inline src/shared/telemetry/defaultTelemetryService.ts:142]: Can we tie this specifically to tests, or can we expose this functionality to release users? It's in a weird state where it's only for tests and it provides no additional benefit for development users (maybe we could drop a breakpoint here? but you don't need the logger for inspection), and could impact performance if lots of metadata is generated.

File: src/shared/telemetry/defaultTelemetryClient.ts [hunk 11]
```diff
@@ -17,14 +17,31 @@ import { ServiceConfigurationOptions } from 'aws-sdk/lib/service'
...
 
+export class DefaultTelemetryClient implements TelemetryClient {
+    private static readonly DEFAULT_IDENTITY_POOL = 'us-east-1:820fd6d1-95c0-4ca4-bffb-3f01d32da842'
+    private static readonly DEFAULT_TELEMETRY_ENDPOINT = 'https://client-telemetry.us-east-1.amazonaws.com'
     private static readonly PRODUCT_NAME = 'AWS Toolkit For VS Code'
 
+    private static initializeConfig(): TelemetryConfiguration {
+        const settings = new DefaultSettingsConfiguration()
+        const identityPool = settings.readDevSetting<string>('aws.dev.telemetryUserPool', 'string', true)
+        const endpoint = settings.readDevSetting<string>('aws.dev.telemetryEndpoint', 'string', true)
```
Comment (bryceitoc9) [inline src/shared/telemetry/defaultTelemetryClient.ts:33]: Is there a reason for this? I'm kinda thinking that if someone wanted to set up their own telemetry endpoint/user pool, it'd be something they'd want settable outside the context of a dev setting.

File: src/shared/telemetry/telemetryLogger.ts (hunk 9 above)
Comment (JadenSimon) [reply src/shared/telemetry/telemetryLogger.ts:62]: Dead code.

File: src/shared/telemetry/defaultTelemetryService.ts (hunk 10 above)
Comment (JadenSimon) [reply src/shared/telemetry/defaultTelemetryService.ts:142]: It'd work fine in dev mode too, though are you thinking of some sort of UI? That could be interesting.

File: src/shared/telemetry/defaultTelemetryClient.ts (hunk 11 above)
Comment (JadenSimon) [reply src/shared/telemetry/defaultTelemetryClient.ts:33]: Oh I just added it since I was working on my own telemetry stack. We already set endpoints for other APIs so it does fit, though honestly it'd be one of the least used settings.

File: src/test/shared/credentials/loginManager.test.ts [hunk 12]
```diff
@@ -66,7 +65,18 @@ describe('LoginManager', async function () {
     })
 
     function assertTelemetry(expected: Parameters<typeof recordAwsValidateCredentials>[0]): void {
```
Comment (justinmk3) [inline src/test/shared/credentials/loginManager.test.ts:67]: should we extract this to testUtil.ts

**Review by justinmk3 — APPROVED** [review]
Exciting!

File: src/shared/telemetry/defaultTelemetryService.ts (hunk 10 above)
Comment (bryceitoc9) [reply src/shared/telemetry/defaultTelemetryService.ts:142]: It currently serves no purpose unless you're running a test against it, which isn't ideal if you're not testing. UI could be interesting (I think JB has something similar), but would probably be low priority since that kind of thing runs heavy in VS Code (easiest thing would just be a command palette command to the logs or something).

File: src/shared/telemetry/telemetryLogger.ts (hunk 8 above)
Comment (JadenSimon) [reply src/shared/telemetry/telemetryLogger.ts:60]: Considering querying for the raw metric is uncommon I'll split it. Overloading on discriminated unions works very well but it's not needed here. The issue with overloading happens when types cannot be discriminated.

File: src/shared/telemetry/telemetryLogger.ts (hunk 7 above)
Comment (JadenSimon) [reply src/shared/telemetry/telemetryLogger.ts:76]: Type checker cannot find the predicate when using an arrow function. Could statically type but eh. A couple `.bind`s won't hurt.

File: src/shared/telemetry/telemetryLogger.ts (hunk 6 above)
Comment (JadenSimon) [reply src/shared/telemetry/telemetryLogger.ts:25]: Probably not. I was using the unions since I did have more fields.

File: src/shared/telemetry/telemetryLogger.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    /** Metric name to look up in the log */
+    readonly metricName: string
+    /** Returns the metric containing the metadata instead of just the metadata (default: false) */
+    readonly returnMetric?: boolean
+    /** Attributes to filter out of the metadata */
+    readonly filters?: string[]
+}
+
+interface MetricQuery extends MetadataQuery {
+    readonly returnMetric: true
```
[the remark this answers is not in the document]
Comment (JadenSimon) [reply src/shared/telemetry/telemetryLogger.ts:18]: They can be eliminated. I had more fields but dropped them for simplicity. Looks like I can cut even more.

File: src/test/shared/credentials/loginManager.test.ts (hunk 12 above)
Comment (JadenSimon) [reply src/test/shared/credentials/loginManager.test.ts:67]: Yeah I'll clean it up and just handle all the edge cases. Unfortunate that the logger won't be too ergonomic on its own (at least for now) but eventually it'll get there.

File: src/shared/telemetry/defaultTelemetryService.ts (hunk 10 above)
Comment (JadenSimon) [reply src/shared/telemetry/defaultTelemetryService.ts:142]: Oh you mean dump the formatted metric to the log. I have that as a commit but haven't pushed it.

**Review by justinmk3 — APPROVED** [review]
